File: java_epi/BuyAndSellStock.java
package java_epi;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by dmorales on 20/09/2016.
 */
public class BuyAndSellStock {

    public static double computeMaxProfit(List<Double> prices) {
        double minPrice = Double.MAX_VALUE; // this stores the minimum price in the list
        double maxProfit = 0.0; // this stores the current max profit per iteration

        for (Double price : prices) {
                // get the max between the current maxProfit and the difference between the current price and the current min price
            double currentPriceAndMinPriceDifference = price - minPrice;
            maxProfit = Math.max(maxProfit, currentPriceAndMinPriceDifference);
            minPrice = Math.min(minPrice, price);
        }

        return maxProfit;

    }

    public static void main(String[] args) {
        List<Double> prices = new ArrayList<>();

        prices.add(100.0);
        prices.add(20.0);
        prices.add(1.0);
        prices.add(80.00);
        prices.add(30.0);
        prices.add(15.0);
        prices.add(101.00);

        System.out.println("max profit: " + computeMaxProfit(prices));

    }

}

File: js_data_structures/BuyAndSellStock.js
/**
 * Created by dmorales on 20/09/2016.
 */

function computeMaxProfit(prices) {
    var minPrice = Number.MAX_VALUE; // this stores the minimum price in the array
    var maxProfit = 0.0; // this stores the current max profit per iteration

    prices.forEach(function (price) {
        // get the max between the current maxProfit and the difference between the current price and the current min price
        var differnceBetweenCurrentPriceAndMinPrice = price - minPrice;
        maxProfit = Math.max(maxProfit, differnceBetweenCurrentPriceAndMinPrice);
        minPrice = Math.min(minPrice, price);
    });

    return maxProfit;

}

var stockPrices = [];

stockPrices.push(10);
stockPrices.push(20);
stockPrices.push(30);
stockPrices.push(10);
stockPrices.push(100);
stockPrices.push(200);
stockPrices.push(0.5);

console.log("Max profit: ", computeMaxProfit(stockPrices));
